Update build-on-mobile.mdx with android result documentation. (#22)
MobilePayDev/MobilePay-Payments-API: docs/guides/app-payments/build-on-mobile.mdx

- You are generally familiar with developing applications on Android.

:::caution Make sure you are compliant with mobile application distribution service rules!
Using app payments for digital sales might not be allowed by some mobile application distribution services (Google Play, App Store, etc.) and might result in your application being removed. Please review the terms of service for the respective mobile application distribution services you use to ensure that you are compliant with their guidelines.
:::

## Create Payment

First, you'll need to create payment. To do this you should call your backend service that is implemented as described [here](/docs/payments-refunds/create-payments). This service call should return a response that contains a deep link needed to launch the MobilePay app.

## Starting MobilePay

MobilePay supports navigation via native Activity Result APIs and deep linking.

<Tabs
  groupId="android-navigation-method"
  defaultValue="result"
  values={[
    { label: 'Result APIs', value: 'result', },
    { label: 'Deep Links', value: 'deep_link', },
  ]
}>
<TabItem value="result">

> Activity Result or `startActivityForResult` APIs use standard android functions and will launch MobilePay in your app process. This is the preferred way when integrating into native android app.

Reviewer: ```suggestion
Activity Result or `startActivityForResult` APIs use standard Android functions and will launch MobilePay in your app process. This is the preferred way when integrating into a native Android app.
```

Let's open MobilePay with the deep link received from your backend.

<Tabs
  groupId="android-language"
  defaultValue="kotlin"
  values={[
    { label: 'Kotlin using Activity Result APIs', value: 'kotlin', },
    { label: 'Kotlin using startActivityForResult', value: 'activity_result_kotlin', },
    { label: 'Java', value: 'java', },
  ]
}>

<TabItem value="kotlin">

> ```kotlin title="Kotlin using Activity Result APIs (Requires AndroidX Activity 1.2.0 or Fragment 1.3.0) "

Reviewer: ```suggestion
```kotlin title="Kotlin using Activity Result APIs (requires AndroidX Activity 1.2.0 (or above) or Fragment 1.3.0 (or above))"
```

private val mobilePayContract = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
  when (it.resultCode) {
    SUCCESS_RESULT_CODE -> {}
    REJECT_RESULT_CODE -> {}
    FAILED_RESULT_CODE -> {}
    UNKNOWN_RESULT_CODE -> {}
  }
}

> // Use the link received from backend to launch MobilePay.

Reviewer: ```suggestion
// Use the link received from your backend to launch MobilePay.
```

fun openMobilePay(deepLink: String) {
  try {
      val intent = Intent(Intent.ACTION_VIEW).apply {
          data = Uri.parse(deepLink)
      }
      mobilePayContract.launch(intent)
  } catch (exception: ActivityNotFoundException) {

>       // Inform the user that MobilePay is not installed.

Reviewer: Are we "allowed" to provide the MP package name here in this documentation? If yes, then we could do:

```suggestion
      // Inform the user that MobilePay is not installed (or even better, direct the user to the package <package> directly in Google Play).
```

Author: added links to play stores.

  }
}

companion object {
    private const val SUCCESS_RESULT_CODE = Activity.RESULT_OK
    private const val REJECT_RESULT_CODE = Activity.RESULT_CANCELED
    private const val FAILED_RESULT_CODE = 1
    private const val UNKNOWN_RESULT_CODE = 2
}
```

</TabItem>

<TabItem value="activity_result_kotlin">

```kotlin title="Kotlin"

> // Use the link received from backend to launch MobilePay.

Reviewer: ```suggestion
// Use the link received from your backend to launch MobilePay.
```

fun openMobilePay(deepLink: String) {
    try {
        val intent = Intent(Intent.ACTION_VIEW).apply {
            data = Uri.parse(deepLink)
        }
        startActivityForResult(intent, MOBILEPAY_PAYMENT_REQUEST_CODE)
    } catch (exception: ActivityNotFoundException) {

>         // Inform the user that MobilePay is not installed.

Reviewer: Same here

    }
}

override fun onActivityResult(
    requestCode: Int,
    resultCode: Int,
    data: Intent?
) {
  super.onActivityResult(requestCode, resultCode, data)
  if (requestCode == MOBILEPAY_PAYMENT_REQUEST_CODE) {
    when (resultCode) {
      SUCCESS_RESULT_CODE -> {}
      REJECT_RESULT_CODE -> {}
      FAILED_RESULT_CODE -> {}
      UNKNOWN_RESULT_CODE -> {}
    }
  }
}

companion object {
  const val MOBILEPAY_PAYMENT_REQUEST_CODE = 1337 // Any request code you choose

  private const val SUCCESS_RESULT_CODE = Activity.RESULT_OK
  private const val REJECT_RESULT_CODE = Activity.RESULT_CANCELED
  private const val FAILED_RESULT_CODE = 1
  private const val UNKNOWN_RESULT_CODE = 2
}
```

</TabItem>

<TabItem value="java">

```java title="Java"
private static final int MOBILEPAY_PAYMENT_REQUEST_CODE = 1337; // Any request code you choose

private static final int SUCCESS_RESULT_CODE = Activity.RESULT_OK;
private static final int REJECT_RESULT_CODE = Activity.RESULT_CANCELED;
private static final int FAILED_RESULT_CODE = 1;
private static final int UNKNOWN_RESULT_CODE = 2;

> // Use the link received from backend to launch MobilePay.

Reviewer: ```suggestion
// Use the link received from your backend to launch MobilePay.
```

void openMobilePay(String deepLink) {
    try {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(deepLink));
        startActivityForResult(intent, MOBILEPAY_PAYMENT_REQUEST_CODE);
    } catch (android.content.ActivityNotFoundException exception) {

>         // Inform the user that MobilePay is not installed.

Reviewer: Same

>     } catch (NullPointerException exception) {

Reviewer: Couldn't we just annotate `deepLink` with NonNull?

Author: I think that annotation requires a dependancy, so I wanted to avoid that.

        // Null deepLink was passed in.
    }
}

@Override
protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    super.onActivityResult(requestCode, resultCode, data);

    if (requestCode == MOBILEPAY_PAYMENT_REQUEST_CODE) {
        if (resultCode == SUCCESS_RESULT_CODE) {
          
        } else if (resultCode == REJECT_RESULT_CODE) {
            
        } else if (resultCode == FAILED_RESULT_CODE) {
            
        } else if (resultCode == UNKNOWN_RESULT_CODE) {
            
        }
    }
}
```
</TabItem>

</Tabs>

## Handling the Result

MobilePay can return these result codes:

>   - <b> SUCCESS_RESULT_CODE</b> (Activity.RESULT_OK) - payment was successful and it is in reserved state.

Reviewer: ```suggestion
  - <b>SUCCESS_RESULT_CODE</b> (Activity.RESULT_OK) - payment was successful and it is in reserved state.
```

  - <b>REJECT_RESULT_CODE</b> (Activity.RESULT_CANCELED) - user rejected the payment.

>   - <b>FAILED_RESULT_CODE</b> (1) - user could not complete the payment due to invalid state of payment or his MobilePay account.

Reviewer: ```suggestion
  - <b>FAILED_RESULT_CODE</b> (1) - user could not complete the payment due to invalid state of payment or the user's MobilePay account.
```

  - <b>UNKNOWN_RESULT_CODE</b> (2) - user attempted to do a payment, but because of network issues we could not determine if it was successful. 

>   (E.g. User tries to do a payment, but it does not succeed due to network issues. Then he decides to reject the payment and again he receives a network error. 

Reviewer: ```suggestion
  (E.g., the user tries to do a payment, but it does not succeed due to network issues. Then the user decides to reject the payment and again receives a network error.
```

>   At this point we cannot leave him blocked in MobilePay app, so we finish with this result code.) 

Reviewer: ```suggestion
  At this point we cannot leave the user blocked in the MobilePay app, so we finish with this result code.)
```

  You should try to capture or at least check the state of these payments as they can be successful.

</TabItem>

<TabItem value="deep_link">

> Deep link approach is used when MobilePay app is launched with `startActivity` function. When using this navigation method the MobilePay app will run in separate process and invoke your app with a deep link after payment flow is finished.

Reviewer: ```suggestion
Deep link approach is used when the MobilePay app is launched with the `startActivity` function. When using this navigation method, the MobilePay app will run in a separate process and invoke your app with a deep link after the payment flow is finished.
```

Also, did you mean startActivityForResult instead of startActivity?

    }
}
```
</TabItem>
</Tabs>

After this, the MobilePay app will be launched and the user will be able to proceed with the payment flow.

## Returning to Your App

There are multiple ways a user can return to your app:

- User approves the payment and is redirected back to your app.
- User cancels the payment and is redirected back to your app.
- User fails to complete the payment due to some kind of error and is redirected back to your app.
- User closes the app before/after reservation using the native app switcher.
- User presses back in the MobilePay Login screen.

All these can be grouped into:

- App switching.
- Redirect.

The following section addresses how each one should be handled.

### App Switching

> At any point user can close MobilePay app and just return to your app. **This can happen even if payment is already reserved!** Since it is unknown if your app was opened after completed payment, you'll need to check payment state each time your app returns to the foreground. For this, you should use your backend service that provides current state of payment (more details [here](/docs/payments-refunds/create-payments)).

Reviewer: ```suggestion
At any point the user can close the MobilePay app and just return to your app. **This can happen even if the payment is already reserved!** Since it is unknown if your app was opened after a completed payment, you'll need to check the payment state each time your app returns to the foreground. For this, you should use your backend service that provides the current state of a payment (more details [here](/docs/payments-refunds/create-payments)).
```

### Redirect from MobilePay

The other scenario is that the MobilePay app tries to automatically navigate back to your app. For this to work you'll need:

- Valid deep link to a place that knows how to proceed with the flow ([official deep linking guide](https://developer.android.com/training/app-links/deep-linking)).
- Provide this deep link during [payment creation](#create-payment).

In the general case, this deep link should open the app screen which handles the "App switching" scenario described above, and just treat it the same way as if the user manually navigated to your app.

:::tip Payment state in your deep link

When redirecting back to your app, MobilePay appends a query parameter `status` to the end of a deep link. It can contain these values:

- successful
- failed
- rejected

> Consider checking this parameter and triggering a capture without checking payment's state in your backend (capture can only succeed if `status` returned value is `successful`).

Reviewer: ```suggestion
Consider checking this parameter and triggering a capture without checking the payment's state in your backend (capture can only succeed if `status` returned value is `successful`).
```

P.S. It's not required for a successful app payments integration.

:::

</TabItem>
</Tabs>
</TabItem>
<TabItem value="ios">

## Prerequisites and Assumptions

This guide assumes:

- You already have backend services integrated with [MobilePay APIs](/docs/guides/app-payments/how-it-works).
- Your app handles a custom URL scheme. See [Apple documentation](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app) for more details.

> - You are generally familiar with developing applications on iOS.

Reviewer: This guide has only mentioned Android stuff so far, so why the mentions about Apple and iOS here suddenly?

Reviewer: Hmm or probably it's because this is the beginning of an iOS section. I just don't see any header about this.

Author: the above `<TabItem value="ios">` separates this into a new tab.
<img width="1680" alt="Screenshot 2021-11-10 at 13 43 19" src="https://user-images.githubusercontent.com/54581781/141107133-3d029f68-6b09-419d-be4c-6466ceeea6d0.png">
